core/src/ballot_box.rs

```diff
@@ -0,0 +1,400 @@
+            msg!("Ballot box account has an invalid owner");
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+        if epoch_snapshot.data_is_empty() {
+            msg!("Ballot box account data is empty");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if expect_writable && !epoch_snapshot.is_writable {
+            msg!("Ballot box account is not writable");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if epoch_snapshot.data.borrow()[0].ne(&Self::DISCRIMINATOR) {
+            msg!("Ballot box account discriminator is invalid");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        if epoch_snapshot
+            .key
+            .ne(&Self::find_program_address(program_id, ncn, ncn_epoch).0)
+        {
+            msg!("Ballot box account is not at the correct PDA");
+            return Err(ProgramError::InvalidAccountData);
+        }
+        Ok(())
+    }
+
+    pub fn slot_consensus_reached(&self) -> u64 {
+        self.slot_consensus_reached.into()
+    }
+
+    pub fn unique_ballots(&self) -> u64 {
+        self.unique_ballots.into()
+    }
+
+    pub fn operators_voted(&self) -> u64 {
+        self.operators_voted.into()
+    }
+
+    pub fn is_consensus_reached(&self) -> bool {
+        self.slot_consensus_reached() == 0
+    }
+
+    pub fn get_winning_ballot(&self) -> Result<Ballot, TipRouterError> {
+        if self.winning_ballot.is_cast() {
+            Ok(self.winning_ballot)
+        } else {
+            Err(TipRouterError::ConsensusNotReached)
+        }
+    }
+
+    fn increment_or_create_ballot_tally(
+        &mut self,
+        ballot: &Ballot,
+        stake_weight: u128,
+    ) -> Result<usize, TipRouterError> {
+        let mut tally_index: usize = 0;
+        for tally in self.ballot_tallies.iter_mut() {
+            if tally.ballot.eq(ballot) {
+                tally.increment_tally(stake_weight)?;
+                return Ok(tally_index);
+            }
+
+            if tally.is_empty() {
+                *tally = BallotTally::new(*ballot, stake_weight);
+
+                self.unique_ballots = PodU64::from(
+                    self.unique_ballots()
+                        .checked_add(1)
+                        .ok_or(TipRouterError::ArithmeticOverflow)?,
+                );
+
+                return Ok(tally_index);
+            }
+
+            tally_index = tally_index
+                .checked_add(1)
+                .ok_or(TipRouterError::ArithmeticOverflow)?;
+        }
+
+        Err(TipRouterError::BallotTallyFull)
+    }
+
+    pub fn cast_vote(
+        &mut self,
+        operator: Pubkey,
+        ballot: Ballot,
+        stake_weight: u128,
+        current_slot: u64,
+    ) -> Result<(), TipRouterError> {
+        let ballot_index = self.increment_or_create_ballot_tally(&ballot, stake_weight)?;
+
+        for vote in self.operator_votes.iter_mut() {
```

Comment:
thoughts on doing ncn operator index change here vs future PR?

Reply:
I've added the `ncn_operator_index` to the operator struct. Keeping it non-indexed for now